# Author Info
---
Name: **Ejaz-ur-Rehman**\
Business Unit Head | Data Analyst | Crystal Tech (Project of MUZHAB Group)\
Faculty Member-Department of Management Sciences | Al Ghazali University\
MS (Finance), PhD Scholar\
Karachi, Pakistan

![Date](https://img.shields.io/badge/Date-01--August--2026-green?logo=google-calendar)
[![Email](https://img.shields.io/badge/Email-ijazfinance%40gmail.com-blue?logo=gmail)](mailto:ijazfinance@gmail.com)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-Ejaz--ur--Rehman-blue?logo=linkedin)](https://www.linkedin.com/in/ejaz-ur-rehman/)
[![GitHub](https://img.shields.io/badge/GitHub-ejazurrehman-black?logo=github)](https://github.com/ejazurrehman)

# Machine Learning and Predictive Analytics for Islamic Banking

## Section-01: Introduction

### 1.1 Background

Machine Learning has become one of the most influential technologies in modern financial services. Islamic banks increasingly rely on data-driven techniques to forecast financial performance, identify growth opportunities, assess operational risks, and support strategic decision-making. Unlike traditional statistical analysis, machine learning algorithms can discover complex relationships among financial indicators and generate predictions from historical data.

The previous notebooks established the foundation for predictive analytics. The first notebook explored the dataset through exploratory data analysis (EDA), while the second notebook transformed the raw data into a cleaner and more informative dataset using feature engineering techniques. This notebook builds upon those results by applying machine learning algorithms to the engineered dataset.

### 1.2 Notebook Objectives

After completing this notebook, students will be able to:
- Understand the workflow of a machine learning project in Islamic banking.
- Convert long-format financial data into a machine learning dataset.
- Prepare predictor and target variables.
- Develop regression models for predicting financial indicators.
- Develop classification models for business decision support.
- Evaluate model performance using appropriate metrics.
- Interpret feature importance and model outputs.
- Compare different machine learning algorithms for Islamic banking applications.

### 1.3 Learning Outcomes

By the end of this notebook, students should be able to:

- Prepare financial datasets for machine learning.
- Differentiate between regression and classification problems.
- Train multiple machine learning models using Scikit-learn.
- Evaluate predictive performance using quantitative metrics.
- Explain the business implications of predictive analytics in Islamic banking.

## Section 02: Import Required Libraries

### 2.1 Purpose

This section imports all Python libraries required throughout the notebook. The libraries support data manipulation, visualization, preprocessing, machine learning, and model evaluation.

In [1]:
# ============================================================
# Machine Learning Libraries for Islamic Banking Analytics
# ============================================================

# Data manipulation
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Machine Learning preprocessing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder

# Regression Models
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

# Classification Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

# Regression Evaluation
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

# Classification Evaluation
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report,
    roc_auc_score,
    roc_curve
)

# Ignore warnings
import warnings
warnings.filterwarnings("ignore")

# Plot style
plt.style.use("ggplot")

print("Libraries imported successfully.")

Libraries imported successfully.


## Section 3: Load the Data Set and Prepratoin for Machine Learning

### 3.1 Load the Feature-Engineered Dataset

In [2]:
# ============================================================
# Section 2.2: Load Feature-Engineered Dataset
# ============================================================

import pandas as pd

# Load the feature-engineered dataset
df = pd.read_csv("Islamic_Banking_Feature_Engineered.csv")

print(f"Dataset Shape: {df.shape}")

display(df.head())

Dataset Shape: (85839, 19)


,Country,Regional Code,Report Type,Time Period,USD Exchange Rate,Indicator Code,Indicator Description,currency,units,Values,Values in USD Millions,Year,Quarter,Month,Month_Name,Year_Quarter,Indicator Category,Frequency,Country_Indicator_Count
0,Bangladesh,SA,Islamic Banking,2013-10-01,77.4,AD06_010_010_010,"(a) agriculture, forestry, hunting and fishing",NC,M,16880.01,218.09,2013,4,10,October,2013-Q4,Other,906,136
1,Bangladesh,SA,Islamic Banking,2013-10-01,77.4,AD07_010_010_010,"(a) agriculture, forestry, hunting and fishing",NC,M,603.10,7.79,2013,4,10,October,2013-Q4,Other,906,136
2,Bangladesh,SA,Islamic Banking Windows,2013-10-01,77.4,AD06_010_010_010,"(a) agriculture, forestry, hunting and fishing",NC,M,641.62,8.29,2013,4,10,October,2013-Q4,Other,906,136
3,Bangladesh,SA,Islamic Banking Windows,2013-10-01,77.4,AD07_010_010_010,"(a) agriculture, forestry, hunting and fishing",NC,M,1.08,0.01,2013,4,10,October,2013-Q4,Other,906,136
4,Bangladesh,SA,Islamic Banking,2014-01-01,77.4,AD06_010_010_010,"(a) agriculture, forestry, hunting and fishing",NC,M,16880.01,218.09,2014,1,1,January,2014-Q1,Other,906,136


### 3.2 Initial Data Inspection

Before preparing the machine learning dataset, it is important to verify the dataset structure, variable types, and missing values.

In [3]:
print(df.info())

display(df.describe())

display(df.isnull().sum())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85839 entries, 0 to 85838
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Country                  85839 non-null  object 
 1   Regional Code            85839 non-null  object 
 2   Report Type              85839 non-null  object 
 3   Time Period              85839 non-null  object 
 4   USD Exchange Rate        85839 non-null  float64
 5   Indicator Code           85839 non-null  object 
 6   Indicator Description    85839 non-null  object 
 7   currency                 68648 non-null  object 
 8   units                    70176 non-null  object 
 9   Values                   73846 non-null  float64
 10  Values in USD Millions   73846 non-null  float64
 11  Year                     85839 non-null  int64  
 12  Quarter                  85839 non-null  int64  
 13  Month                    85839 non-null  int64  
 14  Month_Name            

,USD Exchange Rate,Values,Values in USD Millions,Year,Quarter,Month,Frequency,Country_Indicator_Count
count,85839.000000,7.384600e+04,73846.000000,85839.000000,85839.000000,85839.000000,85839.000000,85839.000000
mean,3136.032175,1.531091e+05,10233.061312,2018.978483,2.516688,5.550065,793.330596,135.400238
std,5887.339808,5.057889e+05,37268.579079,3.386500,1.122881,3.368643,453.818384,1.031853
min,3.210000,-2.584355e+04,-4236.860000,2013.000000,1.000000,1.000000,44.000000,134.000000
25%,4.050000,2.793000e+01,4.894551,2016.000000,2.000000,4.000000,453.000000,135.000000
50%,83.900000,3.457575e+03,182.635000,2019.000000,3.000000,7.000000,862.000000,135.000000
75%,280.190000,6.622165e+04,2574.160000,2022.000000,4.000000,10.000000,906.000000,136.000000
max,16421.000000,9.327270e+06,486387.800000,2025.000000,4.000000,10.000000,2265.000000,137.000000


Country                        0
Regional Code                  0
Report Type                    0
Time Period                    0
USD Exchange Rate              0
Indicator Code                 0
Indicator Description          0
currency                   17191
units                      15663
Values                     11993
Values in USD Millions     11993
Year                           0
Quarter                        0
Month                          0
Month_Name                     0
Year_Quarter                   0
Indicator Category             0
Frequency                      0
Country_Indicator_Count        0
dtype: int64

### 3.4 Select the Analytical Population

#### Purpose

The IFSB dataset contains multiple reporting populations, including
Islamic Banking and Islamic Banking Windows.

To ensure consistency throughout the machine learning analysis,
this notebook focuses exclusively on observations reported under
**Islamic Banking**.

Filtering the dataset at this stage ensures that each Country–Year–Quarter
combination represents a single banking population and prevents duplicate
observations during the pivot operation.

In [4]:
# ============================================================
# Section 2.4: Select the Analytical Population
# ============================================================

print("Available Report Types\n")

display(df["Report Type"].value_counts())

# Keep only Islamic Banking observations
df = df[
    df["Report Type"] == "Islamic Banking"
].copy()

print(f"\nDataset Shape after Filtering: {df.shape}")

Available Report Types



Report Type
Islamic Banking Windows    43070
Islamic Banking            42769
Name: count, dtype: int64


Dataset Shape after Filtering: (42769, 19)


In [5]:
print(df["units"].value_counts())

units
M      27892
B       7840
n.a      908
Name: count, dtype: int64


In [8]:
# Keep only Million units
df = df[df["units"] == "M"].copy()

### 3.5 Create Unique Feature Names

In [7]:
df["Feature_Name"] = (
    df["Indicator Code"]
    + " | "
    + df["Indicator Description"]
)

### 3.6 Create the Machine Learning Dataset

In [9]:
df_ml = (
    df.pivot_table(
        index=[
            "Country",
            "Year",
            "Quarter",
            "Year_Quarter"
        ],
        columns="Feature_Name",
        values="Values",
        aggfunc="first"
    )
    .reset_index()
)

### 3.7 Verify Dataset Integrity

In [10]:
print(f"Duplicate Columns: {df_ml.columns.duplicated().sum()}")

assert df_ml.columns.duplicated().sum() == 0

print("✓ No duplicate columns found.")

Duplicate Columns: 0
✓ No duplicate columns found.


In [11]:
print(f"Machine Learning Dataset Shape: {df_ml.shape}")

display(df_ml.head())

Machine Learning Dataset Shape: (178, 169)


Feature_Name,Country,Year,Quarter,Year_Quarter,AD01_010 | Income distributed to IAH,AD01_020 | Total income from assets funded by PSIA,AD02_010 | Off-balance sheet items,AD02_020 | Total assets,AD03_010 | FX funding,AD03_020 | Total funding,...,ST07_010_0120 | Ijārah/ Ijārah Muntahia Bittamlīk/ Hire Purchase/ HPSM,ST07_010_020 | Commodity Murābahah / Tawwaruq,ST07_010_030 | Salam,ST07_010_040 | Istisnā`,ST07_010_050 | Ijārah/Ijārah Muntahia Bittamlīk,ST07_010_060 | Muḍārabah,ST07_010_070 | Mushārakah,ST07_010_080 | Diminishing Mushārakah,ST07_010_090 | Wakālah,ST08 | Assets held by domestic systemically important Islamic banks
0,Bangladesh,2013,4,2013-Q4,94346.00,120907.02,282717.45,1361995.55,23736.52,1244232.75,...,210202.13,NaN,9545.15,2781.57,NaN,19785.80,21287.40,NaN,10865.08,NaN
1,Bangladesh,2014,1,2014-Q1,94346.00,120907.02,282717.45,1361995.55,23736.52,1244232.75,...,210202.13,34.023333,9545.15,2781.57,NaN,19785.80,21287.40,1076.046667,10865.08,NaN
2,Bangladesh,2014,2,2014-Q2,96773.63,123794.31,326845.54,1499745.42,33644.32,1411437.34,...,230007.55,7567.003333,9338.39,3205.13,NaN,21438.28,19847.14,1344.720000,10698.95,NaN
3,Bangladesh,2014,3,2014-Q3,96773.63,123794.31,326845.54,1499745.42,33644.32,1411437.34,...,230007.55,7558.363333,9338.39,3205.13,NaN,21438.28,19847.14,1259.183333,10698.95,NaN
4,Bangladesh,2014,4,2014-Q4,96159.34,127986.72,306686.91,1631365.14,33170.24,1502584.82,...,251538.41,29.983333,10528.26,31319.74,NaN,27542.23,21734.08,1029.496667,9778.71,NaN


### 3.8 Define the Final Machine Learning Dataset

## Section 4: Data Transformation for Machine Learning

### 4.1 Purpose

The feature-engineered dataset currently exists in long format, where each row represents a single financial indicator for a country during a specific reporting period. While this format is suitable for data storage and exploratory analysis, it is not appropriate for most machine learning algorithms.

Machine learning models require a wide-format dataset, where each row represents a unique observation and each financial indicator becomes a separate feature (predictor). Therefore, the first step in predictive analytics is to reshape the dataset using a pivot operation.

The transformation performed in this section will:

- Convert the dataset from long format to wide format.
- Create one row for each country and reporting period.
- Convert financial indicators into independent predictor variables.
- Produce a machine learning-ready dataset for regression and classification.

### 4.2 Understanding the Transformation
Long Format (Current Dataset)
| Country  | Year_Quarter | Indicator Description  | Values |
| -------- | ------------ | ---------------------- | -----: |
| Pakistan | 2024-Q1      | Total Assets           |  8,750 |
| Pakistan | 2024-Q1      | Total Deposits         |  6,540 |
| Pakistan | 2024-Q1      | Return on Assets       |   1.12 |
| Pakistan | 2024-Q1      | Capital Adequacy Ratio |  16.30 |

Each financial indicator is stored in a separate row.

Wide Format (Machine Learning Dataset)
| Country  | Year_Quarter | Total Assets | Total Deposits |  ROA |   CAR |
| -------- | ------------ | -----------: | -------------: | ---: | ----: |
| Pakistan | 2024-Q1      |         8750 |           6540 | 1.12 | 16.30 |

Each row now represents a complete financial profile of a country for a specific reporting period.

This structure is suitable for machine learning because all predictor variables are available within a single observation.

### 4.3 Create the Pivot Table

### 4.4 Inspect the New Dataset

#### 4.4.1 Display First Five Records

#### 4.4.2 Check Dataset Dimensions

#### 4.4.3 Check Variable Types

#### 4.4.4 Display Variable Names

### 4.5 Verify the Transformation

The following checks confirm that the transformation has been performed correctly.

#### 4.5.1 Check for Duplicate Observations

#### 4.5.2 Check Missing Values

#### 4.5.3 Summary Statistics

### 4.6 Save the Machine Learning Dataset

The transformed dataset will be used throughout the remainder of the notebook.

## Interpretation

The dataset has now been transformed into a machine learning-ready structure. Each row represents a unique country during a specific reporting period, while each column represents a financial indicator that can serve as a predictor in subsequent models.

The pivot transformation has introduced missing values because not every country reports every financial indicator in every reporting period. Before developing predictive models, these missing values must be carefully examined and treated. The next section focuses on data preparation for machine learning, including missing value analysis, feature selection, and data quality assessment.

## Section 5: Data Preparation for Machine Learning

### 5.1 Purpose

The machine learning dataset created in the previous section contains one observation for each country and reporting period. However, the transformation from long format to wide format introduced missing values because not every financial indicator is reported by every country during every reporting period.

Before building predictive models, it is essential to evaluate the quality of the transformed dataset. This section focuses on identifying missing values, selecting an appropriate treatment strategy, removing variables with excessive missingness, and verifying that the final dataset is suitable for machine learning.

By the end of this section, the dataset will be fully prepared for feature selection and predictive modeling.

### 5.2 Overview of Missing Values

The first step is to identify variables containing missing observations.

### Interpretation

The output displays the number of missing observations for each variable. These missing values are expected because several financial indicators are reported only by selected countries or during specific reporting periods. Therefore, missing values at this stage do not necessarily indicate data quality issues but rather differences in reporting practices across jurisdictions.

### 5.3 Missing Value Percentage

Absolute counts alone do not fully describe the extent of missingness. Calculating the percentage of missing observations provides a more meaningful basis for deciding whether a variable should be retained, imputed, or removed.

### 5.4 Visualizing Missing Values

A graphical representation helps identify variables with substantial missingness.

### 5.5 Missing Value Treatment Strategy

Before applying any treatment, a clear decision rule is required. The following framework will be used throughout this notebook.
| Missing Percentage | Treatment Strategy                                    |
| -----------------: | ----------------------------------------------------- |
|                 0% | Retain variable                                       |
|       Less than 5% | Median imputation                                     |
|          5% to 30% | Evaluate business importance, then impute if retained |
|         30% to 70% | Retain only if financially significant                |
|   Greater than 70% | Remove from the dataset                               |

**NOTE**: This framework balances statistical quality with the practical relevance of financial indicators.

### 5.6 Remove Variables with Excessive Missing Values

Variables containing more than 70% missing observations contribute limited information and may adversely affect model performance. Such variables will therefore be removed.

### 5.7 Impute Remaining Missing Values

After removing variables with excessive missingness, the remaining numerical variables are imputed using the median. Median imputation is appropriate because it is robust to extreme values that are common in financial datasets.

### 5.8 Verify Data Quality

The final verification confirms that the dataset is complete and ready for predictive modeling.

#### 5.8.1 Display Dataset Information

#### 5.8.2 Display Dataset Shape

### 5.9 Save the Clean Machine Learning Dataset

The cleaned dataset will be used throughout the remainder of the notebook.

## Interpretation

The machine learning dataset has now been cleaned and prepared for predictive analytics. Variables with excessive missing values were removed, while the remaining numerical variables were completed using median imputation. These preprocessing steps improve data quality and ensure that machine learning algorithms can be trained without interruptions caused by incomplete observations.

The resulting dataset provides a reliable foundation for model development while preserving the most informative financial indicators.

## Section 6: Feature Selection and Target Variable Identification

### 6.1 Purpose

Feature selection is a critical step in the machine learning pipeline. Although the cleaned dataset contains numerous financial indicators, not every variable should be included in predictive modeling. Some variables serve only as identifiers, while others may introduce redundancy or target leakage, leading to overly optimistic model performance.

This section aims to identify informative predictor variables and establish suitable target variables for both regression and classification. The resulting feature set will provide a reliable foundation for developing predictive models in the following sections.

### 6.2 Understanding Predictor and Target Variables

In supervised machine learning, the dataset is divided into two components:

- Predictor variables (X): Independent variables used by the model to learn patterns.
- Target variable (y): The outcome that the model attempts to predict.

For this notebook:

- Regression: Predict a continuous financial indicator.
- Classification: Predict a categorical banking class derived from financial data.

### 6.3 Review Available Variables

Before selecting predictors, review the variables available in the cleaned dataset.

### 6.4 Remove Identifier Variables

Several variables uniquely identify an observation but contain no predictive information. These variables should not be used as model inputs.

> Note: These variables will remain in the dataset for reporting and interpretation but will be excluded from the feature matrix (X).

### 6.5 Detect Highly Correlated Variables

Highly correlated variables may provide duplicate information and increase multicollinearity, particularly in regression models.

### 6.6 Identify Highly Correlated Pairs

To simplify interpretation, identify variable pairs with an absolute correlation greater than 0.90.

### Interpretation

A high correlation between two predictors indicates that they contain similar information. During model development, only one of the variables may be retained to reduce redundancy and improve interpretability. However, removal should also consider financial relevance rather than relying solely on statistical thresholds.

### 6.7 Data Integrity Check

#### 6.7.1 Identify Duplicate Columns

#### 6.6.2 Investigate Duplicate Columns

Before removing them, compare whether the duplicate columns contain identical values.

#### 6.7.3 Remove Duplicate Columns

If the duplicates are confirmed to be identical, retain the first occurrence.

#### 6.7.4 Verify Data Integrity

### 6.7 Selection of Target Variables
Purpose

The objective of supervised machine learning is to predict an outcome variable using a set of explanatory variables. Therefore, selecting an appropriate target variable is one of the most important decisions in model development.

This notebook demonstrates both regression and classification using the same Islamic banking dataset. Consequently, two target variables will be prepared:

- A continuous target for regression.
- A categorical target for classification.

Before selecting these targets, it is important to avoid target leakage, where predictor variables contain direct information about the target variable, resulting in unrealistically high prediction accuracy.

#### 6.7.1 Define the Regression Target

Following the data integrity checks, Total assets is selected as the regression target because it is a continuous financial indicator that reflects the overall size of the Islamic banking sector and is widely used in banking analysis.

## Section 7: Regression Modeling

### 7.1 Purpose of Regression Modeling

Regression modeling is one of the most widely used supervised machine learning techniques for predicting continuous numerical values. Unlike classification models, which assign observations to predefined categories, regression models estimate the actual magnitude of a target variable based on a set of explanatory features.

In the context of Islamic banking, predicting a bank's Total Assets is a valuable analytical task because it provides insights into institutional growth, financial stability, and operational performance. By learning the relationships between financial indicators and total assets, regression models can assist researchers, analysts, and decision-makers in forecasting future asset values and identifying the factors that contribute most to bank performance.

In this notebook, Total Assets has been selected as the regression target variable. The feature-engineered dataset prepared in Notebook 02 will serve as the input for building predictive models. The regression workflow begins with a baseline Linear Regression model, followed by more advanced machine learning algorithms in the next section.

Throughout this process, the notebook follows the standard machine learning pipeline, including data preparation, train-test splitting, feature scaling, model training, model evaluation, visualization, and interpretation of results. This structured approach promotes reproducibility and reflects best practices in financial data analytics.

# 7.2 Prepare the Regression Dataset

## Purpose

Before developing a regression model, the predictor variables (**X**) and the target variable (**y**) must be clearly defined.

In this notebook, the regression target is **AD02_020 | Total assets**, which represents the Total Assets indicator selected from the IFSB dataset.

During data validation, several variables representing alternative forms of Total Assets were identified. Two of these variables (**CP11_020 | Total assets** and **CP13_020 | Total assets**) were found to be perfectly correlated with the selected target variable. Including them as predictors would introduce perfect multicollinearity and result in data leakage.

Therefore, these redundant variables are removed from the predictor matrix.

Additionally, identifier variables such as Country and Year_Quarter are preserved separately for reporting purposes but are excluded from model training because they do not represent numerical financial characteristics.

At the end of this section, the machine learning dataset is divided into:

- **X_reg** : Predictor variables
- **y_reg** : Regression target variable

These datasets will be used throughout the regression modelling process.

### Interpretation

The regression dataset has been successfully prepared.

The predictor matrix (**X_reg**) contains only numerical financial indicators suitable for machine learning.

The regression target (**y_reg**) contains the selected Total Assets variable.

Perfectly correlated variables were excluded to prevent multicollinearity and target leakage.

Furthermore, all identifier variables have been removed from the predictor matrix while being preserved separately for future prediction reports.

The dataset is now ready to be partitioned into training and testing subsets.

### Next Step

With the regression dataset prepared, the next step is to divide the data into training and testing subsets. This enables the model to learn from one portion of the data while being evaluated on previously unseen observations.

The train-test split will be implemented in **Section 7.3**.

# 7.3 Train-Test Split

## Purpose

Machine learning models must be evaluated on data that were not used during model training. To achieve this, the regression dataset is divided into two subsets:

- **Training Set:** Used to train the regression model.
- **Testing Set:** Used to evaluate the predictive performance on unseen observations.

In this notebook, 80% of the observations are allocated to the training dataset, while the remaining 20% are reserved for testing.

To ensure reproducibility, a fixed random state is used during the data partitioning process.

Additionally, the observation identifiers (**Country** and **Year_Quarter**) are split alongside the predictor and target variables. These identifiers will later be merged with model predictions to produce country-wise prediction reports and performance summaries.

### Interpretation

The regression dataset has been successfully divided into training and testing subsets.

The training dataset will be used to estimate the parameters of the regression model, while the testing dataset will remain unseen during training and will be used to evaluate the model's predictive performance.

The observation identifiers (**Country** and **Year_Quarter**) have been preserved separately. This ensures that every prediction generated later in the notebook can be linked back to its corresponding country and reporting period, enabling the creation of detailed prediction reports and country-wise performance summaries.

### Next Step

Machine learning algorithms often perform better when numerical features are measured on a comparable scale. Therefore, the next section standardizes the predictor variables using feature scaling before fitting the regression model.

# 7.4 Feature Scaling

## Purpose

Machine learning algorithms often perform better when numerical variables are measured on a comparable scale. Since the financial indicators in this dataset are expressed in different units and magnitudes, feature scaling is applied before model training.

In this notebook, **StandardScaler** is used to standardize the predictor variables. The scaler transforms each feature by subtracting its mean and dividing by its standard deviation, resulting in variables with approximately zero mean and unit variance.

The scaler is fitted **only on the training dataset** to prevent information from the testing dataset from influencing the model. The same fitted scaler is then applied to the testing dataset.

The target variable (**y_reg**) is **not scaled** because Linear Regression predicts the original values of the dependent variable.

### Interpretation

The predictor variables have now been standardized using the statistics of the training dataset. This preprocessing step ensures that differences in measurement scales do not disproportionately influence machine learning algorithms.

By fitting the scaler only on the training data and applying the learned transformation to the testing data, the notebook follows a key machine learning best practice that prevents data leakage and preserves the integrity of model evaluation.

The scaled datasets are now ready for model training.

>Note: The original target variable (y_train_reg and y_test_reg) is not scaled because Linear Regression predicts the target directly in its original units (Total Assets), making the results easier to interpret.

### 7.5 Linear Regression

#### Purpose

Linear Regression is one of the most widely used supervised machine learning algorithms for predicting continuous numerical values. It models the relationship between a dependent variable and one or more independent variables by fitting a linear equation to the observed data.

In this notebook, Linear Regression serves as the baseline regression model for predicting the **Total Assets** of Islamic banks. Although more sophisticated algorithms will be introduced later, establishing a baseline allows us to evaluate whether advanced machine learning techniques provide meaningful improvements in predictive performance.

The model is trained using the standardized predictor variables prepared in the previous section and will generate predictions for both the training and testing datasets.

#### Mathematical Representation

For multiple predictor variables, the Linear Regression model is expressed as:

$$
Y=\beta_0+\beta_1X_1+\beta_2X_2+\cdots+\beta_nX_n+\varepsilon
$$



where

- **Y** = Predicted Total Assets
- **β₀** = Intercept
- **β₁ … βₙ** = Regression coefficients
- **X₁ … Xₙ** = Predictor variables
- **ε** = Random error

#### Workflow

The Linear Regression model will be developed using the following steps:

1. Initialize the Linear Regression model.
2. Train the model using the training dataset.
3. Generate predictions for both training and testing data.
4. Examine the model intercept and regression coefficients.
5. Prepare the model for performance evaluation in the next section.

#### Regression Coefficients

Each regression coefficient measures the expected change in the predicted value of **Total Assets** associated with a one-unit increase in the corresponding predictor variable, assuming all other variables remain constant.

Positive coefficients indicate a direct relationship with Total Assets, while negative coefficients indicate an inverse relationship.

### Standardized Coefficient Analysis

### Interpretation

The Linear Regression model has successfully learned the relationship between the engineered financial indicators and the Total Assets of Islamic banks.

The regression coefficients provide insight into the relative influence of each predictor variable. Variables with larger absolute coefficient values generally have a stronger impact on the predicted Total Assets. However, coefficient interpretation should be made cautiously when predictor variables exhibit multicollinearity.

The model has now been trained and is ready for quantitative evaluation using regression performance metrics. In the next section, we will assess prediction accuracy using MAE, MSE, RMSE, and the coefficient of determination (R²).

### 7.6 Model Evaluation

### Purpose

Model evaluation is a critical stage in the machine learning workflow. After training the Linear Regression model, its predictive performance must be assessed using data that were not used during model training.

To obtain a comprehensive understanding of model performance, evaluation metrics are calculated for both the training and testing datasets. Comparing these results helps determine whether the model generalizes well or suffers from overfitting or underfitting.

In this section, four commonly used regression performance metrics are employed:

- Mean Absolute Error (MAE)
- Mean Squared Error (MSE)
- Root Mean Squared Error (RMSE)
- Coefficient of Determination (R²)